[Lim Wan Ning] iP

wanninglim opened the pull request:
# Ponyo

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Ponyo frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/wanninglim/ip/releases/tag/v0.1).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

wanninglim committed:
Add increment Level 1

wanninglim committed:
Add increment Level 2

wanninglim committed:
Add increment Level 3

wanninglim committed:
Add increment Level 4

wanninglim committed:
Add increment A-TextUiTesting

wanninglim committed:
Add increment Level-5

wanninglim committed:
Add increment Level-6

wanninglim committed:
Clean up code

wanninglim committed:
Add increment Level-7

wanninglim committed:
Duke: Clean up code

wanninglim committed:
Add increment Level-8

wanninglim committed:
Duke: Clean up code

wanninglim committed:
Merge branch 'branch-Level-8'

wanninglim committed:
Refactor code to use OOP

wanninglim committed:
Add organisation into packages

wanninglim committed:
Add JUnit tests: AddCommand, TaskList

wanninglim committed:
Add manifest.mf

wanninglim committed:
Add increment Level-9

wanninglim committed:
Add JavaDocs

wanninglim committed:
Clean up code

wanninglim committed:
Merge branch 'Level-9' into master

wanninglim committed:
Merge branch 'A-CodingStandard'

wanninglim committed:
Clean up code to match previous changes

wanninglim committed:
Merge branch 'A-JavaDoc'

wanninglim committed:
Add JavaDocs for FindCommand

wanninglim committed:
Add Gradle support

wanninglim committed:
Update code to match checkstyle

wanninglim committed:
Merge branch 'branch-A-Gradle' into master

wanninglim committed:
Merge branch 'branch-A-CheckStyle'

wanninglim committed:
Ponyo: Clean up code

charltonator commented on src/main/java/ponyo/commands/AddCommand.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Adds a new task to the task list.
+ */
+public class AddCommand extends Command {
```
Would it be better if the class was named CommandAdder? I think AddCommand might sound like a method name :)

Anonymxtrix replied:
No actually I think that `CommandAdder` will distort the meaning of the class. This class is meant to represent the command to add a task into the task list. As such, either `Add` or `AddCommand` should be fine. Any thoughts?

charltonator commented on src/main/java/ponyo/commands/AddCommand.java:
```diff
@@ -0,0 +1,99 @@
+     * @return The string representation of the date in MMM d yyyy
+     * @throws DateTimeParseException Thrown when the inputDate is of wrong format.
+     */
+    public String dateFormatter(String inputDate, int slashIndex) throws DateTimeParseException {
```
Would it be better to name this method formatDate so it follows coding standards better? 💯

wanninglim replied:
That's true! Thanks for the feedback! :-)

charltonator commented on src/main/java/ponyo/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Deletes a task identified using its index in the task list.
+ */
+public class DeleteCommand extends Command {
```
Same as AddCommand maybe the class name could be CommandDeleter

Anonymxtrix replied:
Refer to the original comment on AddCommand for my reply

charltonator commented on src/main/java/ponyo/commands/ListCommand.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Displays the list of stored tasks.
+ */
+public class ListCommand extends Command {
```
Same as the other classes that inherit from command maybe you could change the naming?

Anonymxtrix replied:
Same reply as the above similar comments :)

charltonator commented on src/main/java/ponyo/data/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @param task the task to be added to the list
+     */
+    public void add(Task task) {
```
Maybe better if the name was clearer such as addTask?

Anonymxtrix replied:
I think its fine since the type of the parameter when mousing over the add method would have already implied that this is dealing with adding a `Task`. What are your thoughts?

charltonator commented on src/main/java/ponyo/storage/Storage.java:
```diff
@@ -0,0 +1,178 @@
+     * @param filePath path to check
+     * @throws IOException if file creation fails
+     */
+    public void fileFolderCheck(String filePath) throws IOException {
```
maybe better if the method was named checkFileFolder?

wanninglim replied:
Thanks for spotting that!

wanninglim committed:
Add basic GUI features for Ponyo

wanninglim committed:
Update GUI features to include Ponyo functionalities

wanninglim committed:
Merge branch 'branch-Level-10' into master

wanninglim committed:
Update files to match coding standards

Anonymxtrix commented on src/main/java/ponyo/Main.java:
```diff
@@ -0,0 +1,32 @@
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
```
Perhaps we can consider using `anchorPane` here instead of `ap`. I get that since the number of lines in this function is short, the meaning of `ap` is pretty obvious, however once the number of lines here get too long or the `Scene scene = new Scene(ap);` line moves too far away, the meaning of `ap` might not be as obvious anymore.

Anonymxtrix commented on src/main/java/ponyo/Ponyo.java:
```diff
@@ -0,0 +1,56 @@
+     * @return Response message by system.
+     */
+    public String[] handleInput(String input) {
+        boolean isExit = false;
```
Maybe `isExiting` or `isExited` since I can't really understand what this `isExit` is used for O.o

wanninglim replied:
Ah, I used `isExit` to symbolise whether a type of command is meant to exit the application! And you're right haha after looking at it now it does seem a little vague! How does `isExitCommand` sound?

Anonymxtrix commented on src/main/java/ponyo/commands/AddCommand.java:
```diff
@@ -0,0 +1,95 @@
+ * Adds a new task to the task list.
+ */
+public class AddCommand extends Command {
+    private final String[] taskToAdd;
```
Since this variable is of type `String[]`, I think it would be nicer to have a plural name like `tasksToAdd` instead. There are a few other similar instances in your pull request (eg. Line 21).

wanninglim replied:
Oops! Must have missed this out when I was refactoring my code haha T_T

Anonymxtrix commented on src/main/java/ponyo/storage/Storage.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @param t task to be written into file
+     */
+    public void fileLineToWrite(Task t) {
```
Since method names should make use of verbs, I suggest using `writeTaskToFile` or `appendTaskToFile` or `addTaskToFile` instead. This is because of two reasons:
- Method names should make use of verbs. 
- Method name tells me you are dealing with "line" when the parameter is actually a `Task`. 

Alternatively, `addToFile` or `writeToFile` etc. may also be good enough since the parameter already implies that we are only working with a `Task`.

wanninglim replied:
Thanks for spotting this :-)

Anonymxtrix commented on src/main/java/ponyo/commands/AddCommand.java:
```diff
@@ -0,0 +1,95 @@
+            }
+        case "deadline":
+            try {
+                int slashIndex = taskToAdd[1].indexOf("/");
```
How about making `"/"` into a final constant since its used in a few places in this method? What are your thoughts?

Anonymxtrix commented on src/main/java/ponyo/commands/Command.java:
```diff
@@ -0,0 +1,23 @@
+     * @return true if the command is meant to exit the program,
+     *         false otherwise.
+     */
+    public abstract boolean isExit();
```
I think I know what your `isExit` is trying to do now. Perhaps `isExitingCommand` might be more descriptive about the use of this method?

wanninglim replied:
Just saw your comment here! That sounds good too 👍

Anonymxtrix commented on src/main/java/ponyo/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,31 @@
+ * Deletes a task identified using its index in the task list.
+ */
+public class DeleteCommand extends Command {
+    private final int taskToDelete;
```
Perhaps `taskIdToDelete` would be better here since we are working with an `int` instead of a `Task` type. Similarly for other `Command` type class fields which are `int`s like the one in `DoneCommand` etc...

wanninglim replied:
Sounds good to me. 👍

Anonymxtrix commented on src/main/java/ponyo/commands/ExitCommand.java:
```diff
@@ -0,0 +1,21 @@
+ * Exits the program.
+ */
+public class ExitCommand extends Command {
+    private static final String MESSAGE_BYE = "\tBye. Hope to see you again soon!";
```
This is probably just me being picky but maybe we can try `BYE_MESSAGE` since it reads better than `MESSAGE_BYE`.

wanninglim replied:
I named the constants such that all messages were prefixed with `MESSAGE` :) but your suggestion definitely sounds better!

Anonymxtrix commented on src/main/java/ponyo/commands/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+
+        if (taskClone.size() != 0) {
```
In cases like this, I love to use `taskClone.isEmpty()` instead since it reads better than `taskClone.size() != 0` although both works nicely.

wanninglim replied:
It completely slipped my mind that there's such a method in ArrayList! Thanks for your input :)

Anonymxtrix commented on src/main/java/ponyo/commands/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public String[] execute(TaskList tasks, Storage storage) {
+        ArrayList<Task> taskClone = new ArrayList<>();
```
Perhaps `taskListClone` or `tasksClone`? Although I think something like `matchingTasks` would do better in terms of explaining what this represents.

Anonymxtrix commented on src/main/java/ponyo/data/task/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by + ")";
```
Nice! Although perhaps we can consider putting `"D"` here in a final constant or variable since its used both in your `toStringInFile` and `toString` methods. Similarly for the rest of the `Task` related classes.

Anonymxtrix commented on src/main/java/ponyo/data/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+        return "[" + this.getStatusIcon() + "] " + this.description;
+    }
+
+    public boolean containsKeyword(String kw) {
```
I think not really necessary but I would suggest using `keyword` instead for the same reason as `ap` above.

Anonymxtrix commented on src/main/java/ponyo/data/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+    protected List<Task> tasks;
+
+    public TaskList() {
+        this.tasks = new ArrayList<>();
```
Not naming related but you could probably make use of `super(new ArrayList<>());` here instead to link this constructor to the one defined in line 16.

Anonymxtrix commented on src/main/java/ponyo/data/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @param task the task to be removed from the list.
+     */
+    public void remove(int task) {
```
Perhaps using `taskId` here might be more indicative of the parameter type instead. Same for the `retrieveTask` method.

Anonymxtrix commented on src/main/java/ponyo/storage/Storage.java:
```diff
@@ -0,0 +1,178 @@
+     * @param filePath path of file
+     * @throws IOException if error reading file
+     */
+    public void readTasksFromFile(ArrayList tasks, String filePath) throws IOException {
```
Also not naming related but I am curious why you decided to take in the `ArrayList tasks` as a parameter instead of either returning a newly created `ArrayList` or returning the raw text from `Stoage` class and parsing it in your `Tasks` or `TaskList` class.

wanninglim replied:
Ah you're right! Made a rookie mistake of mutating my arguments in the process. 👎  Thanks! :)

Anonymxtrix commented on src/main/java/ponyo/ui/Ui.java:
```diff
@@ -0,0 +1,70 @@
+        show(msg);
+    }
+
+    /**
+     * Shows messages to user.
+     *
+     * @param messages the messages to be printed
+     * @return an array of strings to be printed in order
+     */
+    public static String[] show(String... messages) {
+        return messages;
+    }
+
+    /**
+     * Shows a message to user.
+     *
+     * @param msg the message to be printed
+     * @return an array of a string to be printed
+     */
+    public static String[] show(String msg) {
+        return new String[] { msg };
+    }
```
Arent these 2 methods the same O.o? Whats the use of the `show` method in line 67?

wanninglim replied:
LOL omg you're right TT thanks!!

Anonymxtrix reviewed (commented):
Nicely done! I like the way you followed all the instructions on the IP project task descriptions to abstract your program using `Command`, `Task`, `Ui`, etc... If you need to clarify any of my comments feel free to tag me (although I tend to not look at my notifs) XD

wanninglim committed:
Refactor code

wanninglim committed:
Add assertions

Add assertions to document important assumptions in the code.

Some assumptions need to hold true and using assertions can help to
validate this on the developer's end.

Let's add some assertions to verify whether our assumptions hold true
during runtime.

wanninglim committed:
Update files to match code standard

Currently, the Storage class consists of a number of duplicate
code that can be extracted into smaller methods.

Let's refactor it to ensure it matches the code standard.

wanninglim committed:
Merge pull request #4 from wanninglim/branch-A-Assertions

wanninglim committed:
Merge branch 'master' into branch-A-CodeQuality

wanninglim committed:
Merge pull request #3 from wanninglim/branch-A-CodeQuality

Update files to match code standard

wanninglim committed:
Set up CI

wanninglim committed:
Merge pull request #5 from wanninglim/branch-A-CI

Set up CI

wanninglim committed:
Add aliases for keywords

wanninglim committed:
Merge pull request #6 from wanninglim/branch-C-FriendlierSyntax

wanninglim committed:
Add better GUI

wanninglim committed:
Update paddings for DialogBox

wanninglim committed:
Update user guide

wanninglim committed:
Set theme jekyll-theme-cayman

wanninglim committed:
Update README.md

wanninglim committed:
Merge branch 'master' of https://github.com/wanninglim/ip

wanninglim committed:
Update tests

wanninglim committed:
Add empty list message

wanninglim committed:
Fix storage issues and add error handling for invalid index

wanninglim committed:
Add styling changes

wanninglim committed:
Update method header comments
